Add Windows support

kaiguo opened the pull request:
<!-- Thanks for submitting a pull request! We appreciate you spending the time to work on these changes. Please follow the template so that the reviewers can easily understand what the code changes affect -->

# Summary

<!--
Explain the **motivation** for making this change: here are some points to help you:

* What issues does the pull request solve? Please tag them so that they will get automatically closed once the PR is merged
* What is the feature? (if applicable)
* How did you implement the solution?
* What areas of the library does it impact?
-->

Adding Windows support.

## Test Plan

<!-- Demonstrate the code is solid. Example: The exact commands you ran and their output, screenshots / videos if the pull request changes UI. -->

### What's required for testing (prerequisites)?
0. [System requirments](https://microsoft.github.io/react-native-windows/docs/rnw-dependencies).
1. Run `yarn start:windows`.
2. Open `example/windows/WebViewWindows.sln` with VisualStudio and run.

## Compatibility

| OS      | Implemented |
| ------- | :---------: |
| Windows | ✅ |

## Checklist

<!-- Check completed item, when applicable, via: [X] -->

- [x] I have tested this on a device and a simulator
- [x] I added the documentation in `README.md`
- [ ] I mentioned this change in `CHANGELOG.md`
- [ ] I updated the typed files (TS and Flow)
- [ ] I added a sample use of the API in the example project (`example/App.js`)

safaiyeh commented:
whaaaaatt @kaiguo this is awesome

safaiyeh commented:
One thing that would help me is having windows part of the example app. I can boot up a windows VM since I am not too familiar with the windows code base. So I can approve with manual testing!

trcoffman commented:
I agree with @safaiyeh, I think that the `example/` app needs to be extended to support windows for this to be merged.

trcoffman commented:
Also if it could be squashed to 1 commit first before it's merged that would be awesome.

trcoffman commented on windows/ReactNativeWebView/ReactWebViewManager.cpp:
```diff
@@ -0,0 +1,173 @@
+        if (auto webView = m_webView.get()) {
+            m_navigationCompletedRevoker = webView.NavigationCompleted(
+                winrt::auto_revoke, [=](auto const& sender, auto const& args) {
+                    OnNavigationCompleted(sender, args);
```
I haven't read enough yet to know that this is actually a bug yet, but capturing `this` in a lambda that is going to be called asynchronously from some other class is often incorrect. If you're 100% sure that `this` will always outlive `webView` then this is OK.

If the lambda outlives `this`, that's a crash right here.

One way to avoid this bug is to use `shared_from_this()`, and the lambda can keep a weak_ptr to the ReactWebViewManager and try to `lock()` it before calling `OnNavigationCompleted`.

trcoffman replied:
It looks like the lifetime of webView is not necessarily properly aligned with the lifetime of ReactWebViewManager. This class only owns a weak_ref to the WebView. I'm not super familiar with the IViewManager interface. Does it guarantee that the ViewManager will outlive the view returned by `CreateView`?

kaiguo replied:
Good catch! I think ViewManager should outlive WebView, I've changed the weak ref to strong ref.

trcoffman commented on windows/ReactNativeWebView/ReactWebViewManager.cpp:
```diff
@@ -0,0 +1,173 @@
+        winrt::IJSValueReader const& commandArgsReader) noexcept {
+        if (auto webView = view.try_as<winrt::WebView>()) {
+            switch (commandId) {
+            case 0: // goForward
```
It would be nice if these were named constants.

kaiguo replied:
Updated.

kaiguo commented:
> One thing that would help me is having windows part of the example app. I can boot up a windows VM since I am not too familiar with the windows code base. So I can approve with manual testing!
@safaiyeh Yeah i'm working on adding a sample app, will ping you when it's ready fro review.

kaiguo commented:
> Also if it could be squashed to 1 commit first before it's merged that would be awesome.

Sure I'll squash to 1 commit before merging.

kaiguo commented:
Hi @safaiyeh, I'm trying to run the example app but got a "Module AppRegistry is not registered callable module" error (tested on master branch with Android emulator, I'm seeing the same error in the windows app I just added too). Is this known?

![image](https://user-images.githubusercontent.com/4424330/75507805-f8103680-5996-11ea-970c-e2270a70d59d.png)

kaiguo commented:
@safaiyeh, I added a windows test app in the example folder, but hit some weird metro issues I wasn't able to resolve :/

You can try `yarn start:windows` and see if that works for you. If it doesn't work, I've created a separate test app [here](https://github.com/kaiguo/react-native-windows-webview-test) with some instruction in readme.

safaiyeh commented:
Thanks @kaiguo I'll set up some time to test tomorrow.

acoates-ms commented on package.json:
```diff
@@ -28,17 +29,19 @@
     "react": "^16.9",
-    "react-native": ">=0.60 <0.62"
+    "react-native": ">=0.60 <0.62",
+    "react-native-windows": "0.61.0-beta.27"
```
This peer dep is too restrictive.  At least make it "^0.61.0-beta.27".   Or once we bring RNW 0.61 out of beta, it could match the react-native dependency version.

kaiguo replied:
Added "^" for now

kaiguo commented:
@safaiyeh did you get a chance to check out the sample app?

namrog84 commented on windows/ReactNativeWebView/ReactWebViewManager.cpp:
```diff
@@ -0,0 +1,267 @@
+    }
+
+    winrt::IAsyncAction ReactWebViewManager::InjectJavaScript(winrt::hstring javaScript) {
+        auto result = co_await m_webView.InvokeScriptAsync(L"eval", { javaScript });
```
Is there any way to do this without using eval?  Since Hermes apparently doesn't support it?

jamonholmgren commented:
We'll want to add Windows to each of the "supported platforms" sections in all of the documentation, where appropriate.

Type | Required | Platform
-- | -- | --
function | No | iOS, macOS, Windows

kaiguo commented:
@jamonholmgren doc should be updated now.

Also @safaiyeh I've fixed the sample app, so you should be able to just do `yarn start:windows` and run the test app. I updated the instructions [here](https://github.com/react-native-community/react-native-webview/blob/e7df9b2c6e321c21fc3d5a7afdd20701a76fc750/docs/Contributing.md#for-windows).

![image](https://user-images.githubusercontent.com/4424330/76559095-b930b500-645b-11ea-843b-c8722d23eb00.png)

kaiguo commented on windows/ReactNativeWebView/ReactNativeWebView.vcxproj:
```diff
@@ -0,0 +1,155 @@
+      <AdditionalOptions>%(AdditionalOptions) /bigobj</AdditionalOptions>
+      <!--Temporarily disable cppwinrt heap enforcement to work around xaml compiler generated std::shared_ptr use -->
+      <AdditionalOptions Condition="'$(CppWinRTHeapEnforcement)'==''">/DWINRT_NO_MAKE_DETECTION %(AdditionalOptions)</AdditionalOptions>
+      <DisableSpecificWarnings>28204</DisableSpecificWarnings>
```
Yeah I think I saw the JSValues issue as well, @vmoroz let me know when your changes are settled? I'll update to the latest package and switch to "AsString()", etc.

kaiguo replied:
Updated : )

kaiguo commented on windows/ReactNativeWebView/ReactNativeWebView.vcxproj:
```diff
@@ -0,0 +1,155 @@
+      <!--Temporarily disable cppwinrt heap enforcement to work around xaml compiler generated std::shared_ptr use -->
+      <AdditionalOptions Condition="'$(CppWinRTHeapEnforcement)'==''">/DWINRT_NO_MAKE_DETECTION %(AdditionalOptions)</AdditionalOptions>
+      <DisableSpecificWarnings>28204</DisableSpecificWarnings>
+      <PreprocessorDefinitions>_WINRT_DLL;%(PreprocessorDefinitions)</PreprocessorDefinitions>
```
I'll make the JSValues changes together with this. Thanks for pointing it out : )

safaiyeh commented:
Apologies for the wait, I'll be testing today @kaiguo. Setting up my Windows VM and all ❤️

safaiyeh commented:
I got it to work on my VM! LGTM, I'll land this right now. Thank you @kaiguo for responding to all the comments and updating accordingly, excited for this.

kaiguo commented:
Thanks @safaiyeh!
